File: kotlin/app/src/main/kotlin/com/hyperbabel/demo/data/Session.kt
/*
 * Session — identity + preferences, persisted across launches.
 *
 * The customer JWT pair lives in EncryptedSharedPreferences (see
 * SecureStore.kt) so ApiClient can rotate it on 401 without
 * round-tripping through Compose state.
 *
 * Single mode: Customer Auth pattern B1 (Firebase Direct Exchange).
 * See https://hyperbabel.com/docs#customer-auth for the full architecture.
 */
package com.hyperbabel.demo.data

import android.content.Context
import android.content.SharedPreferences
import com.hyperbabel.demo.BuildConfig
import com.hyperbabel.demo.api.FirebaseAuthService
import com.hyperbabel.demo.api.FirebaseExchangeResult

object Session {
    private const val FILE_NAME = "hb_session"

    private const val KEY_USER_ID  = "hb_user_id"
    private const val KEY_USERNAME = "hb_user_name"
    private const val KEY_LANG     = "hb_lang"
    private const val KEY_API_URL  = "hb_api_url"

    @Volatile private var prefs: SharedPreferences? = null

    /** Must be called from MainActivity before any field is read. */
    fun init(context: Context) {
        if (prefs != null) return
        synchronized(this) {
            if (prefs != null) return
            prefs = context.applicationContext
                .getSharedPreferences(FILE_NAME, Context.MODE_PRIVATE)
        }
    }

    private val store: SharedPreferences
        get() = prefs ?: error(
            "Session.init(context) was not called — call it from MainActivity.onCreate()",
        )

    var userId: String
        get() = store.getString(KEY_USER_ID, "") ?: ""
        private set(value) { store.edit().putString(KEY_USER_ID, value).apply() }

    var displayName: String
        get() = store.getString(KEY_USERNAME, "") ?: ""
        private set(value) { store.edit().putString(KEY_USERNAME, value).apply() }

    var preferredLangCd: String
        get() = store.getString(KEY_LANG, "en") ?: "en"
        private set(value) { store.edit().putString(KEY_LANG, value).apply() }

    var apiUrl: String
        get() = store.getString(KEY_API_URL, BuildConfig.HB_API_URL) ?: BuildConfig.HB_API_URL
        set(value) {
            val trimmed = value.trim()
            store.edit().putString(KEY_API_URL, trimmed.ifEmpty { BuildConfig.HB_API_URL }).apply()
        }

    /** True iff we have an identity AND a customer JWT in secure storage. */
    val isSignedIn: Boolean
        get() = userId.isNotBlank() &&
            !SecureStore.read(SecureStore.KEY_ACCESS_TOKEN).isNullOrBlank()

    /** Persist the result of a Firebase Direct Exchange. */
    fun persist(
        result: FirebaseExchangeResult,
        fallbackDisplayName: String?,
        langCode: String,
    ) {
        SecureStore.write(SecureStore.KEY_ACCESS_TOKEN,  result.access_token)
        SecureStore.write(SecureStore.KEY_REFRESH_TOKEN, result.refresh_token)
        SecureStore.writeLong(SecureStore.KEY_EXPIRES_AT, result.expires_at)

        val trimmed = fallbackDisplayName?.trim().orEmpty()
        val resolvedName = if (trimmed.isNotEmpty()) trimmed
            else result.external_user_id.take(8)
        val resolvedLang = result.preferred_lang_cd ?: langCode

        userId          = result.external_user_id
        displayName     = resolvedName
        preferredLangCd = resolvedLang
    }

    fun updateLang(langCode: String) {
        preferredLangCd = langCode
    }

    fun signOut() {
        SecureStore.clearAll()
        FirebaseAuthService.signOut()
        userId = ""
        displayName = ""
        // Keep lang + apiUrl across sessions — they aren't identity-bound.
    }
}

File: swift/HyperBabelDemo/Models/Session.swift
/*
 * Session — identity + preferences kept in @Published state for SwiftUI,
 * backed by UserDefaults for restart-survival. The customer JWT pair
 * lives in Keychain (see SecureStore.swift) so ApiClient can rotate it
 * without round-tripping through SwiftUI.
 *
 * Single mode: Customer Auth pattern B1 (Firebase Direct Exchange).
 * See https://hyperbabel.com/docs#customer-auth for the full architecture.
 */
import Foundation
import Combine

final class Session: ObservableObject {
    static let shared = Session()

    @Published var userId: String
    @Published var displayName: String
    @Published var preferredLangCd: String
    @Published var apiUrl: String

    private let defaults = UserDefaults.standard
    private enum Keys {
        static let userId   = "hb_user_id"
        static let display  = "hb_user_name"
        static let lang     = "hb_lang"
        static let apiUrl   = "hb_api_url"
    }

    private init() {
        self.userId          = defaults.string(forKey: Keys.userId)  ?? ""
        self.displayName     = defaults.string(forKey: Keys.display) ?? ""
        self.preferredLangCd = defaults.string(forKey: Keys.lang)    ?? "en"
        self.apiUrl          = defaults.string(forKey: Keys.apiUrl)
            ?? ApiClient.defaultApiUrl
    }

    /// True iff we have an identity AND a customer JWT in Keychain.
    var isSignedIn: Bool {
        !userId.isEmpty && SecureStore.read(.accessToken) != nil
    }

    // ── Persist the result of a Firebase Direct Exchange ───────────────

    func persist(_ result: FirebaseExchangeResult, fallbackDisplayName: String?, langCode: String) {
        SecureStore.write(.accessToken,  result.access_token)
        SecureStore.write(.refreshToken, result.refresh_token)
        SecureStore.writeInt(.expiresAt, result.expires_at)

        let resolvedName = (fallbackDisplayName?.trimmingCharacters(in: .whitespaces).isEmpty == false)
            ? fallbackDisplayName!.trimmingCharacters(in: .whitespaces)
            : String(result.external_user_id.prefix(8))
        let resolvedLang = result.preferred_lang_cd ?? langCode

        userId          = result.external_user_id
        displayName     = resolvedName
        preferredLangCd = resolvedLang

        defaults.set(result.external_user_id, forKey: Keys.userId)
        defaults.set(resolvedName,            forKey: Keys.display)
        defaults.set(resolvedLang,            forKey: Keys.lang)
    }

    func updateLang(_ langCode: String) {
        preferredLangCd = langCode
        defaults.set(langCode, forKey: Keys.lang)
    }

    func setApiUrl(_ url: String) {
        let trimmed = url.trimmingCharacters(in: .whitespaces)
        apiUrl = trimmed.isEmpty ? ApiClient.defaultApiUrl : trimmed
        defaults.set(apiUrl, forKey: Keys.apiUrl)
    }

    func signOut() {
        SecureStore.clearAll()
        FirebaseAuthService.signOut()
        userId = ""
        displayName = ""
        defaults.removeObject(forKey: Keys.userId)
        defaults.removeObject(forKey: Keys.display)
        // Keep lang + apiUrl across sessions — they aren't identity-bound.
    }
}
